runtime/src/bank.rs

```diff
@@ -12550,4 +12572,54 @@ pub(crate) mod tests {
             0
         );
     }
+
+    #[test]
+    fn test_clean_dropped_unrooted_unfrozen_banks() {
+        //! Test that dropped unrooted, unfrozen banks are cleaned up properly
+        //!
+        //! slot 0:       bank0 (rooted)
+        //!               /   \
+        //! slot 1:      /   bank1 (unrooted, unfrozen, and dropped)
+        //!             /
+        //! slot 2:  bank2 (rooted)
+        //!
+        //! In the scenario above, when `clean_accounts()` is called on bank2, the keys that exist
+        //! _only_ in bank1 should be cleaned up, since those keys are unreachable.
+        //
+        solana_logger::setup();
+
+        let (genesis_config, mint_keypair) = create_genesis_config(100);
+        let bank0 = Arc::new(Bank::new(&genesis_config));
+
+        let collector = Pubkey::new_unique();
+        let pubkey1 = Pubkey::new_unique();
+        let pubkey2 = Pubkey::new_unique();
+
+        bank0.transfer(2, &mint_keypair, &pubkey2).unwrap();
+        bank0.freeze();
+
+        let slot = 1;
+        let bank1 = Bank::new_from_parent(&bank0, &collector, slot);
+        bank1.transfer(3, &mint_keypair, &pubkey1).unwrap();
+        // bank1 is not frozen on purpose
+
+        let slot = slot + 1;
+        let bank2 = Bank::new_from_parent(&bank0, &collector, slot);
+        bank2.transfer(4, &mint_keypair, &pubkey2).unwrap();
+        bank2.freeze(); // the freeze here is not strictly necessary, but more for illustration
+        bank2.squash();
+
+        drop(bank1);
+
+        bank2.clean_accounts(false);
+        assert_eq!(
+            bank2
+                .rc
+                .accounts
+                .accounts_db
+                .accounts_index
+                .ref_count_from_storage(&pubkey1),
+            0
+        );
+    }
```

Comment:
Nice, I had a few additional test cases in mind, interested in your thoughts if you think they're necessary!

1) An additional case as well that if there was a `pubkey3` touched in *both* (current test case is separate pubkeys being touched) `bank1` and `bank2`, that the reference from the unfrozen `bank1` is cleaned up as well

2) If there was a zero-lamport key `K` in *only* the unfrozen `bank1`, that it is accurately cleaned up.

3)  If there was a zero-lamport key  `K` in *both* the unfrozen `bank1`, and the frozen `bank2` that it is accurately cleaned up.

Reply:
@carllin I like these additional tests. Would it be OK to address them in a subsequent PR, or would you like them in this one?

Reply:
I think it makes sense to combine them under this same PR for tracking purposes and historical reference, but definitely not essential.

Reply:
@carllin Gotcha! I've added the tests now. Let me know if they look like what you were expecting!